Add Benchmark command line reference (#4173)
opensearch-project/documentation-website: _benchmark/commands/execute-test.md

---
layout: default
title: execute-test
nav_order: 65
parent: Command reference
---

> # execute_test

Reviewer: "execute_test" --> "execute-test" as of 0.5.0

- `--results-file`: When provided a file path, writes the compare results to the file indicated.
- `--show-in-results`: Determines whether or not to include the comparison in the results file.

### Distributions

The following options set which version OpenSearch and OpenSearch plugins the benchmark test uses.

- `--distribution-version`: Downloads the specified OpenSearch distribution based on version number. For a list of released OpenSearch versions, see [Version history](https://opensearch.org/docs/version-history/).
- `--distribution-repository`: Defines the repository from where the OpenSearch distribution should be downloaded. Default is `release`.
- `--revision`: Defines the current source code revision for which to run a benchmark test. Default is `current`.
   - `current`: Uses the source tree's current revision based on your OpenSearch distribution. 
   - `latest`: Fetches the latest revision from the main branch of the source tree.
   - You can also a time stamp or commit ID from the source tree. When using a timestamp, specify `@ts` where "ts" is a valid ISO 8601 timestamp, for example `@2013-07-27T10:37:00Z`.

> -  `--opensearch-plugins`: Defines which [OpenSearch plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) to install. By default, not plugins are installed.

Reviewer: "not plugins" --> "no plugins"

- `--plugin-params:` Defines a comma-separated list of key:value pairs that are injected verbatim into all plugins as variables.
- `--runtime-jdk`: The major version of JDK to use.
- `--client-options`: Defines a comma-separated lists of clients to use. All options are passed to the OpenSearch Python client. Default is `timeout:60`.

### Cluster

> The following options set which cluster options. 

Reviewer: Super nit: "The following options et cluster options"

- `--provision-config-path`: Defines the path to the `--provision-config-instance` and any OpenSearch plugin configurations to use.
- `--provision-config-revision`: Defines a specific Git revision in the `provision-config` that OpenSearch Benchmark should use.
- `--provision-config-instance`: Defines the `--provision-config-instance` to use. You can see possible configuration instances using the command `opensearch-benchmark list provision-config-instances`. 
- `--provision-config-instance-params`: A comma-separated list of key:value pairs injected verbatim as variables for the `provision-config-instance`.

### Workload

The following options determine which workload is used to run the test.

- `--workload-repository`: Defines the repository from where OpenSearch Benchmark loads workloads.
- `--workload-path`: Defines the path to a downloaded or custom workload.
- `--workload-revision`: Defines a specific revision from the workload source tree that OpenSearch Benchmark should use.
- `--workload`: Defines the workload to use based on the workload's name. You can find a list of preloaded workloads using `opensearch-benchmark list workloads`.

### Test procedures

> The following options define what test procedures the test uses and which tasks are contained inside the procedure.

Reviewer: Could switch `tasks` to `operations`?

- `--test-execution-id`: Defines a unique ID for this test run.
- `--test-procedure`: Defines a test procedure to use. You can find a list of test procedures using `opensearch-benchmark list test-procedures`.
- `--include-tasks`: Defines a comma-separated list of test procedure tasks to run. By default, all tasks listed in a test procedure array are run.
- `--exclude-tasks`: Defines a comma-separated list of test procedure tasks not to run.
- `--enable-assertions`: Enables assertion checks for tasks. Default is `false`.

### Pipelines

> The `--pipeline` option selects a pipeline to run.

Reviewer 2: Is there a command to list the type of pipelines as well?
If yes, can we add that command here?

### Telemetry

The following options enable telemetry devices on OpenSearch Benchmark.

- `--telemetry`: Enables the provided telemetry devices when the devices are provided using a comma-separated list. You can find a list of possible telemetry devices by using `opensearch-benchmark list telemetry`.
- `--telemetry-params`: Defines a comma-separated list of key:value pairs that are injected verbatim into the telemetry devices as parameters.

### Errors

The following options set how OpenSearch Benchmark handles errors when running tests.

- `--on-error`: Controls how OpenSearch Benchmark responds to errors. Default is `continue`.
  - `continue`: Continues to run the test despite the error.
  - `abort`: Aborts the test when an error occurs.
- `--preserve-install`: Keeps the Benchmark candidate and its index. Default is `false`.

> - `--kill-running-processes`: When set to `true`, stops any OpenSearch processes currently running and allows OpenSearch Benchmark to continue to run. Default is `false`.

Reviewer: Should be "stops any OpenSearch Benchmark processes"

|                 Min Throughput |     painless_dynamic |   2.53292 |  ops/s |
|              Median Throughput |     painless_dynamic |   2.53813 |  ops/s |
|                 Max Throughput |     painless_dynamic |   2.54401 |  ops/s |
|      50.0th percentile latency |     painless_dynamic |    172208 |     ms |
|      90.0th percentile latency |     painless_dynamic |    310401 |     ms |
|      99.0th percentile latency |     painless_dynamic |    341341 |     ms |
|      99.9th percentile latency |     painless_dynamic |    344404 |     ms |
|       100th percentile latency |     painless_dynamic |    344754 |     ms |
| 50.0th percentile service time |     painless_dynamic |    393.02 |     ms |
| 90.0th percentile service time |     painless_dynamic |   407.579 |     ms |
| 99.0th percentile service time |     painless_dynamic |   430.806 |     ms |
| 99.9th percentile service time |     painless_dynamic |   457.352 |     ms |
|  100th percentile service time |     painless_dynamic |   459.474 |     ms |

----------------------------------
[INFO] SUCCESS (took 2634 seconds)
----------------------------------
```

## Options

> Use the following options to customize the `execute-test` command for your use case. Options in this section are categorized by their use-case.

Reviewer 3: use case. no hyphen necessary.

> Use the following options to customize the `execute-test` command for your use case. Options in this section are categorized by their use-case.

Author: ```suggestion
Use the following options to customize the `execute-test` command for your use case. Options in this section are categorized by their use case.
```

## General settings

The following options shape how each test runs and how results are output.

- `--test-mode`: Runs the given workload in test mode, which is useful when checking a workload for errors.

> - `--user-tag`: Defines user-specific key-value pairs to be used in metric record as meta information, for example, `intention:baseline-ticket-12345`.

Reviewer 3: yes, it looks like "key-value pair" is the preferred formatting across the web. (with hyphen)

- `--results-format`: Defines the output format for the command line results, either `markdown` or `csv`. Default is `markdown`.
- `--results-number-align`: Defines the column number alignment for when the `compare` command outputs results. Default is `right`.

> - `--results-file`: When provided a file path, writes the compare results to the file indicated.

Reviewer 3: ```suggestion
- `--results-file`: When provided a file path, writes the compare results to the file indicated in the path.
```

- `--show-in-results`: Determines whether or not to include the comparison in the results file.

### Distributions

The following options set which version OpenSearch and OpenSearch plugins the benchmark test uses.

- `--distribution-version`: Downloads the specified OpenSearch distribution based on version number. For a list of released OpenSearch versions, see [Version history](https://opensearch.org/docs/version-history/).
- `--distribution-repository`: Defines the repository from where the OpenSearch distribution should be downloaded. Default is `release`.

> - `--revision`: Defines the current source code revision for which to run a benchmark test. Default is `current`.

Reviewer 3: ```suggestion
- `--revision`: Defines the current source code revision to use for running a benchmark test. Default is `current`.
```

   - `current`: Uses the source tree's current revision based on your OpenSearch distribution. 
   - `latest`: Fetches the latest revision from the main branch of the source tree.

>    - You can also a time stamp or commit ID from the source tree. When using a timestamp, specify `@ts` where "ts" is a valid ISO 8601 timestamp, for example `@2013-07-27T10:37:00Z`.

Reviewer 3: ```suggestion
   - You can also use a time stamp or commit ID from the source tree. When using a timestamp, specify `@ts` where "ts" is a valid ISO 8601 timestamp, for example `@2013-07-27T10:37:00Z`.
```
Does this description belong with line 106?

-  `--opensearch-plugins`: Defines which [OpenSearch plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) to install. By default, no plugins are installed.

> - `--plugin-params:` Defines a comma-separated list of key:value pairs that are injected verbatim into all plugins as variables.

Reviewer 3: key-value pairs

- `--runtime-jdk`: The major version of JDK to use.

> - `--client-options`: Defines a comma-separated lists of clients to use. All options are passed to the OpenSearch Python client. Default is `timeout:60`.

Reviewer 3: ```suggestion
- `--client-options`: Defines a comma-separated list of clients to use. All options are passed to the OpenSearch Python client. Default is `timeout:60`.
```

### Cluster

> The following options relate to the target cluster of the benchmark.

Reviewer 3: Is the intention that more options will be added at some time in the future? You may still want to be accurate and say "following options relates to ..."

> The following options relate to the target cluster of the benchmark.

Author: ```suggestion
The following option relates to the target cluster of the benchmark.
```

> - `--target-hosts`: Defines a comma-separated list of host:port pairs which should be targeted if using the pipeline `benchmark-only`. Default is `localhost:9200`.

Reviewer 3: host-port pairs?

> - `--target-hosts`: Defines a comma-separated list of host:port pairs which should be targeted if using the pipeline `benchmark-only`. Default is `localhost:9200`.

Author: ```suggestion
- `--target-hosts`: Defines a comma-separated list of host-port pairs which should be targeted if using the pipeline `benchmark-only`. Default is `localhost:9200`.
```

### Distributed workload generation

The following options help those who want to use multiple hosts to generate load to the benchmark cluster.

- `--load-worker-coordinator-hosts`: Defines a comma-separated list of hosts that coordinate loads. Default is `localhost`.

> - `--enable-worker-coordinator-profiling`: Enables an analysis of the performance of call in OpenSearch Benchmark's worker coordinator. Default is `false`.

Reviewer 3: ```suggestion
- `--enable-worker-coordinator-profiling`: Enables an analysis of the performance of a call in OpenSearch Benchmark's worker coordinator. Default is `false`.
```

> - `--enable-worker-coordinator-profiling`: Enables an analysis of the performance of call in OpenSearch Benchmark's worker coordinator. Default is `false`.

Author: ```suggestion
- `--enable-worker-coordinator-profiling`: Enables an analysis of the performance of OpenSearch Benchmark's worker coordinator. Default is `false`.
```

### Provisioning

The following options help customize how OpenSearch Benchmark provisions OpenSearch and workloads.

- `--provision-config-repository`: Defines the repository from which OpenSearch Benchmark loads `provision-configs` and `provision-config-instances`. 
- `--provision-config-path`: Defines the path to the `--provision-config-instance` and any OpenSearch plugin configurations to use.
- `--provision-config-revision`: Defines a specific Git revision in the `provision-config` that OpenSearch Benchmark should use.
- `--provision-config-instance`: Defines the `--provision-config-instance` to use. You can see possible configuration instances using the command `opensearch-benchmark list provision-config-instances`. 

> - `--provision-config-instance-params`: A comma-separated list of key:value pairs injected verbatim as variables for the `provision-config-instance`.

Reviewer 3: key-value pairs

> - `--provision-config-instance-params`: A comma-separated list of key:value pairs injected verbatim as variables for the `provision-config-instance`.

Author: ```suggestion
- `--provision-config-instance-params`: A comma-separated list of key-value pairs injected verbatim as variables for the `provision-config-instance`.
```

### Test procedures

The following options define what test procedures the test uses and which operations are contained inside the procedure.

- `--test-execution-id`: Defines a unique ID for this test run.
- `--test-procedure`: Defines a test procedure to use. You can find a list of test procedures using `opensearch-benchmark list test-procedures`.
- `--include-tasks`: Defines a comma-separated list of test procedure tasks to run. By default, all tasks listed in a test procedure array are run.
- `--exclude-tasks`: Defines a comma-separated list of test procedure tasks not to run.
- `--enable-assertions`: Enables assertion checks for tasks. Default is `false`.

### Pipelines

The `--pipeline` option selects a pipeline to run. You can find a list of pipelines supported by OpenSearch Benchmark by running `opensearch-benchmark list pipelines`.

### Telemetry

The following options enable telemetry devices on OpenSearch Benchmark.

- `--telemetry`: Enables the provided telemetry devices when the devices are provided using a comma-separated list. You can find a list of possible telemetry devices by using `opensearch-benchmark list telemetry`.

> - `--telemetry-params`: Defines a comma-separated list of key:value pairs that are injected verbatim into the telemetry devices as parameters.

Reviewer 3: key-value

> - `--telemetry-params`: Defines a comma-separated list of key:value pairs that are injected verbatim into the telemetry devices as parameters.

Author: ```suggestion
- `--telemetry-params`: Defines a comma-separated list of key-value pairs that are injected verbatim into the telemetry devices as parameters.
```
